sdk/eventgrid/azure-messaging-eventgrid/src/main/java/com/azure/messaging/eventgrid/CloudEvent.java

```diff
@@ -40,22 +40,35 @@
 
     private static final ClientLogger logger = new ClientLogger(CloudEvent.class);
 
-    private boolean parsed = false;
-
-    private static final JsonSerializer deserializer = new JacksonJsonSerializerBuilder()
-        .serializer(new JacksonAdapter().serializer() // this is a workaround to get the FlatteningDeserializer
-            .registerModule(new JavaTimeModule())) // probably also change this to DateTimeDeserializer when/if it
-        .build();                                  // becomes public in core
+    /**
+     * Create an instance of CloudEvent. The source and type are required fields to publish.
+     * @param source a URI identifying the origin of the event. It can't be null or empty.
+     * @param type   the type of event, e.g. "Contoso.Items.ItemReceived". It can't be null or empty.
+     * @param data the payload of this event. Set to null if your event doesn't have the data payload.
+     *             It will be serialized as a String if it's a String, or application/json if it's not a String.
+     */
+    public CloudEvent(String source, String type, Object data) {
```

Comment:
Can `CloudEvent` have no data ? If `data` is optional, why it is provided in constructor ? User can call `setData()` if they need to. constructor are normally used for required parameters.

Also Add `@throws`  clause  , since it will throw `IllegalArgumentException`.  After that  you do not need to additionally  explain `It can't be null or empty` .

Reply:
Yes a CloudEvent can have no data. This sounds weird but the [specification](https://github.com/cloudevents/spec/blob/v1.0/spec.md#event-data) says it's optional. 
We still put it in the constructor because we believe 99% of the time a CloudEvent will have data.